runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,399 @@
+        // This is the first ProcessElement call for this element/restriction.
+        // The element and restriction are available in c.element().
+        WindowedValue<KV<InputT, RestrictionT>> windowedValue =
+            Iterables.getOnlyElement(c.element().elementsIterable());
+        element = windowedValue.withValue(windowedValue.getValue().getKey());
+        restriction = windowedValue.getValue().getValue();
+      } else {
+        // This is not the first ProcessElement call for this element/restriction - rather,
+        // this is a timer firing, so we need to fetch the element and restriction from state.
+        element = elementState.read();
+      }
+
+      final TrackerT tracker = invoker.invokeNewTracker(restriction);
+      @SuppressWarnings("unchecked")
+      final RestrictionT[] residual = (RestrictionT[]) new Object[1];
+      // TODO: Only let the call run for a limited amount of time, rather than simply
+      // producing a limited amount of output.
+      DoFn.ProcessContinuation cont =
+          invoker.invokeProcessElement(
+              makeContext(c, element, tracker, residual, internals), wrapTracker(tracker));
+
+      if (cont == null) {
+        // All work for this element/restriction is completed. Clear state and release hold.
+        elementState.clear();
+        restrictionState.clear();
+        holdState.clear();
+        return;
+      }
+      if (residual[0] == null) {
+        // This means the call completed unsolicited, and the context produced by makeContext()
+        // did not take a checkpoint. Take one now.
+        residual[0] = checkNotNull(tracker.checkpoint());
+      }
+
+      // Save state for resuming.
+      if (wasFirstCall) {
+        elementState.write(element);
+      }
+      restrictionState.write(residual[0]);
+      if (cont.getFutureOutputWatermark() != null) {
+        holdState.add(cont.getFutureOutputWatermark());
+      }
+      // Set a timer to continue processing this element.
+      timerInternals.setTimer(
+          TimerInternals.TimerData.of(
+              StateNamespaces.global(),
+              Instant.now().plus(cont.getResumeDelay()),
+              TimeDomain.PROCESSING_TIME));
+    }
+
+    private DoFn<InputT, OutputT>.ProcessContext makeContext(
+        final ProcessContext baseContext,
+        final WindowedValue<InputT> element,
+        final TrackerT tracker,
+        final RestrictionT[] residualRestrictionHolder,
+        final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+            internals) {
+      return fn.new ProcessContext() {
+        // Commit at least once every 10k output records.  This keeps the watermark advancing
+        // smoothly, and ensures that not too much work will have to be reprocessed in the event of
+        // a crash.
+        // TODO: Also commit at least once every N seconds (runner-specific parameter).
+        private static final int MAX_OUTPUTS_PER_BUNDLE = 10000;
```

Comment:
Seems like 10k should be a runner parameter. For a streamier runner, that may be too much.